build/build.mk

```diff
@@ -87,23 +87,35 @@ define generatePlatformBuildTarget
 build-$(3)-$(1)-$(2):
   $(eval BINS_OUT_DIR_$(1)_$(2) := $(OUT_DIR)/$(1)_$(2)/$(BUILDTYPE_DIR))
 	@echo "$(ARROW) Building $(3) on $(1)/$(2) to $(BINS_OUT_DIR_$(1)_$(2))/$(3)$(BINARY_EXT)"
-	GOOS=$(1) GOARCH=$(2) go build \
+	cd $(4) && GOOS=$(1) GOARCH=$(2) go build \
 		-v \
 		-gcflags $(GCFLAGS) \
-		-ldflags=$(LDFLAGS) \
-		-o $(BINS_OUT_DIR_$(1)_$(2))/$(3)$(BINARY_EXT) \
-		$(4)/
+		-ldflags $(LDFLAGS) \
+		-o $(CURDIR)/$(BINS_OUT_DIR_$(1)_$(2))/$(3)$(BINARY_EXT)
 endef
 
 # defines a target for each binary
 GOOSES := darwin linux windows
 GOARCHES := amd64 arm arm64
-BINARIES := docgen rad applications-rp ucpd controller
-$(foreach ITEM,$(BINARIES),$(eval $(call generateBuildTarget,$(ITEM),./cmd/$(ITEM))))
-$(foreach ARCH,$(GOARCHES),$(foreach OS,$(GOOSES),$(foreach ITEM,$(BINARIES),$(eval $(call generatePlatformBuildTarget,$(OS),$(ARCH),$(ITEM),./cmd/$(ITEM))))))
+BINARIES := docgen:./cmd/docgen \
+	rad:./cmd/rad \
+	applications-rp:./cmd/applications-rp \
+	ucpd:./cmd/ucpd \
+	controller:./cmd/controller \
+	testrp:./test/testrp \
+	magpiego:./test/magpiego
+
+define parseBinary
+$(eval NAME := $(shell echo $(1) | cut -d: -f1))
+$(eval ENTRYPOINT := $(shell echo $(1) | cut -d: -f2))
+endef
+
+$(foreach ITEM,$(BINARIES),$(eval $(call parseBinary,$(ITEM)) $(call generateBuildTarget,$(NAME),$(ENTRYPOINT))))
+
+$(foreach ARCH,$(GOARCHES),$(foreach OS,$(GOOSES),$(foreach ITEM,$(BINARIES),$(eval $(call parseBinary,$(ITEM)) $(call generatePlatformBuildTarget,$(OS),$(ARCH),$(NAME),$(ENTRYPOINT))))))
 
 # list of 'outputs' to build for all binaries
-BINARY_TARGETS:=$(foreach ITEM,$(BINARIES),build-$(ITEM))
+BINARY_TARGETS := $(foreach ITEM,$(BINARIES),$(eval $(call parseBinary,$(ITEM))) build-$(NAME))
```

Comment:
And this aggregates all of the build targets for the current architecture right?

Reply:
Do we need the equivalent for the "platform" targets?

Reply:
The next line creates the necessary targets for different platforms for all the items in the list.